slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -329,6 +332,109 @@ void SLAMServiceImpl::SetUpMapBuilder() {
 }
 
 std::string SLAMServiceImpl::GetLatestJpegMapString(bool add_pose_marker) {
+    cartographer::io::PaintSubmapSlicesResult painted_slices =
+        GetLatestPaintedMapSlices();
+    if (add_pose_marker) {
+        PaintMarker(painted_slices);
+    }
+
+    auto image = io::Image(std::move(painted_slices.surface));
+    return image.WriteJpegToString(50);
+}
+
+bool SLAMServiceImpl::GetLatestPointCloudMapString(std::string &pointcloud) {
+    cartographer::io::PaintSubmapSlicesResult painted_slices =
+        GetLatestPaintedMapSlices();
+    auto painted_surface = painted_slices.surface.get();
+    int width = cairo_image_surface_get_width(painted_surface);
+    int height = cairo_image_surface_get_height(painted_surface);
+
+    // total number of bytes, 32 bits per pixel
+    size_t size_data = width * height * 4;
+    auto data = cairo_image_surface_get_data(painted_surface);
+
+    std::vector<unsigned char> data_vect(data, data + size_data);
+
+    std::stringbuf data_buffer;
+
+    int num_points = 0;
+
+    // Reduce resolution based off number of pixels. Output is pixels to skip
+    int scaleFactor = 4;
+    int scale = size_data / 100000.f / 32.f / scaleFactor;
+
+    // Loop to filter unwanted data and reduce resolution
+    for (int i = 0; i < size_data; i = i + scale * 4) {
+        int rgb = 0;
+        rgb = rgb | ((int)data_vect[i + 0] << 16);
+        rgb = rgb | ((int)data_vect[i + 1] << 8);
+        rgb = rgb | ((int)data_vect[i + 2] << 0);
+
+        // skip pixels that are not in our map(black/past walls)
+        //  this value represents [102,102,102]
+        if (rgb == 6710886) continue;
+
+        // Determine probability based off color pixel
+        int prob = ViamColorToProbability((int)data_vect[i + 2]);
```

Comment:
[minor] this could potentially be moved up as well to avoid extra calculations being performed

Reply:
let me know what you think on positioning now, unsure of this or the RGB check should go first now that I check each channel first

Reply:
Ah I removed the comment I had stating that the format was 32 bit RGBA. The alpha channel(`data_vect[i + 3]`) is not useful in this instance. Will add a comment on that